connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/ProtobufSerializer.scala

```diff
@@ -0,0 +1,264 @@
+        }
+      } catch {
+        case ise: IncompatibleSchemaException =>
+          throw new IncompatibleSchemaException(
+            s"Cannot convert SQL type ${rootCatalystType.sql} to Protobuf type " +
+              s"${rootDescriptor.getName}.",
+            ise)
+      }
+    if (nullable) { (data: Any) =>
+      if (data == null) {
+        null
+      } else {
+        baseConverter.apply(data)
+      }
+    } else {
+      baseConverter
+    }
+  }
+
+  private type Converter = (SpecializedGetters, Int) => Any
+
+  private def newConverter(
+      catalystType: DataType,
+      fieldDescriptor: FieldDescriptor,
+      catalystPath: Seq[String],
+      protoPath: Seq[String]): Converter = {
+    val errorPrefix = s"Cannot convert SQL ${toFieldStr(catalystPath)} " +
+      s"to Protobuf ${toFieldStr(protoPath)} because "
+    (catalystType, fieldDescriptor.getJavaType) match {
+      case (NullType, _) =>
+        (getter, ordinal) => null
+      case (BooleanType, BOOLEAN) =>
+        (getter, ordinal) => getter.getBoolean(ordinal)
+      case (ByteType, INT) =>
+        (getter, ordinal) => getter.getByte(ordinal).toInt
+      case (ShortType, INT) =>
+        (getter, ordinal) => getter.getShort(ordinal).toInt
+      case (IntegerType, INT) =>
+        (getter, ordinal) => {
+          getter.getInt(ordinal)
+        }
+      case (LongType, LONG) =>
+        (getter, ordinal) => getter.getLong(ordinal)
+      case (FloatType, FLOAT) =>
+        (getter, ordinal) => getter.getFloat(ordinal)
+      case (DoubleType, DOUBLE) =>
+        (getter, ordinal) => getter.getDouble(ordinal)
+      case (StringType, ENUM) =>
+        val enumSymbols: Set[String] =
+          fieldDescriptor.getEnumType.getValues.asScala.map(e => e.toString).toSet
+        (getter, ordinal) =>
+          val data = getter.getUTF8String(ordinal).toString
+          if (!enumSymbols.contains(data)) {
+            throw new IncompatibleSchemaException(
+              errorPrefix +
+                s""""$data" cannot be written since it's not defined in enum """ +
+                enumSymbols.mkString("\"", "\", \"", "\""))
+          }
+          fieldDescriptor.getEnumType.findValueByName(data)
+      case (StringType, STRING) =>
+        (getter, ordinal) => {
+          String.valueOf(getter.getUTF8String(ordinal))
+        }
+
+      case (BinaryType, BYTE_STRING) =>
+        (getter, ordinal) => getter.getBinary(ordinal)
+
+      case (DateType, INT) =>
+        (getter, ordinal) => getter.getInt(ordinal)
+
+      case (TimestampType, LONG) =>
+        fieldDescriptor.getContainingType match {
+          // For backward compatibility, if the Protobuf type is Long and it is not logical type
+          // (the `null` case), output the timestamp value as with millisecond precision.
+          case null => (getter, ordinal) => DateTimeUtils.microsToMillis(getter.getLong(ordinal))
```

Comment:
Shall we support the logical type Timestamp?
https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#timestamp

Reply:
@gengliangwang Except for [Scalar Value Types](https://developers.google.com/protocol-buffers/docs/proto3#scalar), all types in protobuf are message types. As a result, there is no exact field-field match between the catalyst and protobuf. I added timestamp support to both from_protobuf and to_protobuf, also a unit test.